mcstas-comps/share/union-lib.c

```diff
@@ -3878,25 +3879,45 @@ int sample_mesh_intersect(double *t,int *num_solutions,double *r,double *v,struc
     //}
     
     // Return all t
+   // printf("\nLooping over solutions");
     int counter2=0;
+    double x_intersect, y_intersect, z_intersect;
     *num_solutions =0;
     for (iter=0; iter < counter ; iter++){
         if (t_intersect[iter] > 0.0){
             t[counter2] = t_intersect[iter];
-            counter2++;
+            x_intersect = t[counter2]*v[0] + x_new;
+		      y_intersect = t[counter2]*v[1] + y_new;
+		      z_intersect = t[counter2]*v[2] + z_new;
+            coordinates = coords_set(x_intersect,y_intersect,z_intersect);
+		      NORM(coordinates.x, coordinates.y, coordinates.z);
+		      nx[counter2] = coordinates.x;
+		      ny[counter2] = coordinates.y;
+		      nz[counter2] = coordinates.z;				
```

Comment:
nx ny and nz are the components of the normal vector for the intersected surface, but they are just provided the coordinates for the intersection instead. This might be inspired by the sphere case, as that happens to work there. In this case we instead need to grab the normal data in the mesh_storage with the appropriate index.